chanbackup/single.go

```diff
@@ -0,0 +1,328 @@
+	// reach the node over in the past, OR we received an incoming
+	// authenticated connection for the stored identity public key.
+	Addresses []net.Addr
+
+	// CsvDelay is the local CSV delay used within the channel. We may need
+	// this value to reconstruct our script to recover the funds on-chain
+	// after a force close.
+	CsvDelay uint16
+
+	// PaymentBasePoint describes how to derive base public that's used to
+	// deriving the key used within the non-delayed pay-to-self output on
+	// the commitment transaction for a node. With this information, we can
+	// re-derive the private key needed to sweep the funds on-chain.
+	PaymentBasePoint keychain.KeyLocator
+
+	// ShaChainRootDesc describes how to derive the private key that was
+	// used as the shachain root for this channel.
+	ShaChainRootDesc keychain.KeyDescriptor
+}
+
+// NewSingle creates a new static channel backup based on an existing open
+// channel. We also pass in the set of addresses that we used in the past to
+// connect to the channel peer.
+func NewSingle(channel *channeldb.OpenChannel,
+	nodeAddrs []net.Addr) Single {
+
+	chanCfg := channel.LocalChanCfg
+
+	// TODO(roasbeef): update after we start to store the KeyLoc for
+	// shachain root
+
+	// We'll need to obtain the shachain root which is derived directly
+	// from a private key in our keychain.
+	var b bytes.Buffer
+	channel.RevocationProducer.Encode(&b) // Can't return an error.
+
+	// Once we have the root, we'll make a public key from it, such that
+	// the backups plaintext don't carry any private information. When we
+	// go to recover, we'll present this in order to derive the private
+	// key.
+	_, shaChainPoint := btcec.PrivKeyFromBytes(btcec.S256(), b.Bytes())
+
+	return Single{
+		ChainHash:        channel.ChainHash,
+		FundingOutpoint:  channel.FundingOutpoint,
+		ShortChannelID:   channel.ShortChannelID,
+		RemoteNodePub:    channel.IdentityPub,
+		Addresses:        nodeAddrs,
+		CsvDelay:         chanCfg.CsvDelay,
+		PaymentBasePoint: chanCfg.PaymentBasePoint.KeyLocator,
+		ShaChainRootDesc: keychain.KeyDescriptor{
+			PubKey: shaChainPoint,
+			KeyLocator: keychain.KeyLocator{
+				Family: keychain.KeyFamilyRevocationRoot,
+			},
+		},
+	}
+}
+
+// Serialize attempts to write out the serialized version of the target
+// StaticChannelBackup into the passed io.Writer.
+func (s *Single) Serialize(w io.Writer) error {
+	// Check to ensure that we'll only attempt to serialize a version that
+	// we're aware of.
+	switch s.Version {
+	case DefaultSingleVersion:
+	default:
+		return fmt.Errorf("unable to serialize w/ unknown "+
+			"version: %v", s.Version)
+	}
+
+	// If the sha chain root has specified a public key (which is
+	// optional), then we'll encode it now.
+	var shaChainPub [33]byte
+	if s.ShaChainRootDesc.PubKey != nil {
+		copy(
+			shaChainPub[:],
+			s.ShaChainRootDesc.PubKey.SerializeCompressed(),
+		)
+	}
+
+	return lnwire.WriteElements(
+		w,
+		byte(s.Version),
```

Comment:
do you think we should also write out the serialized size of the `Single` for each version, VLB (version-length-bytes)? if we try to restore a multi backup containing versions we don't understand, we may not be able to parse anything after an unknown version bc we don't know how big it is

Reply:
Hadn't thought of that, updated to write out the length as a `uint16` before the raw SCB bytes.